lib/Transforms/LayoutOptimization/LayoutOptimization.cpp

```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
+            if (!matrixType) return std::nullopt;
+            return matrixType.getShape();
+          })
+      .Case<linalg::VecmatOp>(
+          [](auto vecmatOp) -> std::optional<ArrayRef<int64_t>> {
+            // Matrix is operand 1: rows x cols
+            auto matrixType =
+                dyn_cast<RankedTensorType>(vecmatOp.getInputs()[1].getType());
+            if (!matrixType) return std::nullopt;
+            return matrixType.getShape();
+          })
+      .Case<linalg::MatmulOp>(
+          [](auto matmulOp) -> std::optional<ArrayRef<int64_t>> {
+            // LHS matrix is operand 0
+            auto lhsType =
+                dyn_cast<RankedTensorType>(matmulOp.getInputs()[0].getType());
+            if (!lhsType) return std::nullopt;
+            return lhsType.getShape();
+          })
+      .Default([](auto) { return std::nullopt; });
+}
+
+// Helper: Build symbolic DAG for a kernel and count rotations
+static Cost computeKernelCostFromDAG(KernelName kernel,
+                                     ArrayRef<int64_t> shape) {
+  using NodeTy = kernel::ArithmeticDagNode<kernel::SymbolicValue>;
+  std::shared_ptr<NodeTy> kernelDag;
+
+  switch (kernel) {
+    case KernelName::Trivial:
+      return 0;
+
+    case KernelName::MatvecDiagonal: {
+      if (shape.size() < 2) return 0;
+      // Create symbolic matrix (rows x cols) and vector (cols)
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      kernel::SymbolicValue symbolicVector({shape[1]});
+      kernelDag =
+          kernel::implementMatvec(kernel, symbolicMatrix, symbolicVector);
+      break;
+    }
+
+    case KernelName::MatvecNaive: {
+      if (shape.size() < 2) return 0;
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      kernel::SymbolicValue symbolicVector({shape[1]});
+      // MatvecNaive not yet implemented in KernelImplementation.h
+      // For now, use diagonal as approximation
+      kernelDag = kernel::implementMatvec(KernelName::MatvecDiagonal,
+                                          symbolicMatrix, symbolicVector);
+      break;
+    }
+
+    case KernelName::VecmatDiagonal: {
+      if (shape.size() < 2) return 0;
+      // VecmatDiagonal: vector (rows) * matrix (rows x cols)
+      // Approximation: similar structure to MatvecDiagonal but transposed
+      kernel::SymbolicValue symbolicVector({shape[0]});
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      // Use same cost as matvec for now (same number of diagonals)
+      kernelDag = kernel::implementMatvec(KernelName::MatvecDiagonal,
+                                          symbolicMatrix, symbolicVector);
+      break;
+    }
+
+    case KernelName::MatmulDiagonal: {
+      if (shape.size() < 2) return 0;
+      // MatmulDiagonal typically implemented via multiple matvecs
+      // Conservative estimate: cost scales with both dimensions
+      return shape[0] + shape[1];
+    }
+
```

Comment:
Let's leave out matmul here since we don't have a kernel for it. I'm surprise it even has a `KernelName`; that is probably my oversight. Perhaps add a line pointing to https://github.com/google/heir/issues/1376 instead, and handle similarly to MatvecNaive

```
// TODO(#1376): evaluate bicyclic matmul kernel cost
```